drivers/regulator/regulator_axp192.c

```diff
@@ -0,0 +1,435 @@
+/*
+ * Copyright (c) 2023 M5Stack
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT xpowers_axp192_regulator
+
+#include <errno.h>
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/regulator.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/linear_range.h>
+#include <zephyr/sys/util.h>
+
+/* X-Powers AXP192 chip ID */
+#define AXP192_CHIP_ID 0x03
+
+/* X-Powers AXP192 register */
+#define AXP192_REG_POWER_STATUS      0x00
+#define AXP192_REG_CHIP_ID           0x03
+#define AXP192_REG_BUCK123_LDO23_CTL 0x12
+#define AXP192_REG_BUCK1_OUT_VOL     0x26
+#define AXP192_REG_BUCK2_OUT_VOL     0x23
+#define AXP192_REG_BUCK3_OUT_VOL     0x27
+#define AXP192_REG_LDO23_OUT_VOL     0x28
+#define AXP192_REG_VBUS_IPS_CTL      0x30
+#define AXP192_REG_CHG_CTL           0x33
+#define AXP192_REG_PER_KEY_CTL       0x36
+#define AXP192_REG_ACIN_IN_VOL       0x56
+#define AXP192_REG_ACIN_IN_CUR       0x58
+#define AXP192_REG_VBUS_IN_VOL       0x5A
+#define AXP192_REG_VBUS_IN_CUR       0x5C
+#define AXP192_REG_INTER_TEMP        0x5E
+#define AXP192_REG_BAT_VOL           0x78
+#define AXP192_REG_BAT_CHG_CUR       0x7A
+#define AXP192_REG_BAT_DCHG_CUR      0x7C
+#define AXP192_REG_GPIO0_MODE_CTL    0x90
+#define AXP192_REG_LDOIO0_OUT_VOL    0x91
+
+/* X-Powers AXP192 buck and ldo output control bit */
+#define AXP192_BUCK1_CTL_BIT BIT(0)
+#define AXP192_BUCK2_CTL_BIT BIT(4)
+#define AXP192_BUCK3_CTL_BIT BIT(1)
+#define AXP192_LDO2_CTL_BIT  BIT(2)
+#define AXP192_LDO3_CTL_BIT  BIT(3)
+
+#define AXP192_GPIO0_LDO_MODE 0x02
+#define AXP192_GPIO0_LOW_MODE 0x05
+
+/* X-Powers AXP192 voltage sources */
+enum axp192_voltage_sources {
+	AXP192_SOURCE_BUCK1,
+	AXP192_SOURCE_BUCK2,
+	AXP192_SOURCE_BUCK3,
+	AXP192_SOURCE_LDO1,
+	AXP192_SOURCE_LDO2,
+	AXP192_SOURCE_LDO3,
+	AXP192_SOURCE_LDOIO0
+};
+
+struct regulator_axp192_pconfig {
+	struct i2c_dt_spec i2c;
+};
+
+struct regulator_axp192_config {
+	struct regulator_common_config common;
+	const struct device *p;
+	uint8_t source;
+};
+
+struct regulator_axp192_data {
+	struct regulator_common_data data;
+};
+
+/* Linear range for output voltage */
+static const struct linear_range buck13_range = LINEAR_RANGE_INIT(700000, 25000, 0U, 112U);
+static const struct linear_range buck2_range = LINEAR_RANGE_INIT(700000, 25000, 0U, 63U);
+static const struct linear_range ldo23_io0_range = LINEAR_RANGE_INIT(1800000, 100000, 0U, 15U);
+
+static int buck_output_enable(const struct device *dev, uint8_t reg_addr, uint8_t ctl_bit,
```

Comment:
buck_output_enable and ldo_output_enable are identical.
You could call i2c_reg_update_byte_dt directly from regulator_axp192_enable